### DISCUSSION TITLE
Troubleshoot SMTP (no email is sent)

### DISCUSSION THREAD

**Discussion by corydodt** [lldap/lldap#941] [Q&A] [answered]
My LLDAP instance does not send password reset emails via my SMTP server, which is Fastmail. I have previously set up outbound SMTP with this service. I have also tested the app password and username successfully with this swaks command (some privacy substitutions below)

```
swaks \
  --to myname@fastmail.com \
  --server smtp.fastmail.com:465 \
  --from myname@example.com \
  --helo "lldap.example.com" \
  --tls-on-connect \
  --auth LOGIN \
  -au myname@example.com \
  -ap xxxxthisismypassword
```

(The above successfully sends an email to me.)

I think this is likely some kind of TLS problem. I am configuring LLDAP with environment variables provided to the container, as follows:

```
LLDAP_SMTP_OPTIONS__ENABLE_PASSWORD_RESET=true
LLDAP_SMTP_OPTIONS__SERVER=smtp.fastmail.com
LLDAP_SMTP_OPTIONS__TLS_REQUIRED=true
# LLDAP_SMTP_OPTIONS__SMTP_ENCRYPTION=TLS
LLDAP_SMTP_OPTIONS__PORT=465
LLDAP_SMTP_OPTIONS__USER=myname@example.com
LLDAP_SMTP_OPTIONS__PASSWORD=xxxxthisismypassword
LLDAP_SMTP_OPTIONS__FROM=myname@example.com
LLDAP_VERBOSE=true
```

I am using LLDAP_SMTP_OPTION__PASSWORD instead of ...PASSWORD_FILE because this env file is processed by a password manager service, so there's no need for a file. Other passwords and secrets in the LLDAP config are working just fine the same way.

The crux of my question:

- Even with LLDAP_VERBOSE, lldap isn't logging any information about what's happening with SMTP. I need to see error messages, timeouts, whatever's available. Are there settings, or tools, that would give me more information?
- Given that the swaks command above works, I am having trouble understanding/parsing how LLDAP uses TLS. Can anyone confirm that my SMTP settings match my swaks settings? I am unsure whether I have chosen the options that match "--tls-on-connect" here, but I have tried other options which aren't working either (e.g. `LLDAP_SMTP_OPTIONS__SMTP_ENCRYPTION`)

Thanks for taking a look!

**Comment by nitnelave** [comment]
Hey! Welcome :)

First things first, can you paste the SMTP config part of your config _as output at the top of a verbose run_? (Of course you can anonymize)
What we're looking for here is a mismatch between your env variables and the parsed config: that would suggest a typo somewhere.

**Reply by corydodt** [reply]
Thanks!

OK, some egg on my face right now: I finally found the correct combination of settings almost as soon as I posted this.

The critical change was that it needs BOTH of these:

```
LLDAP_SMTP_OPTIONS__TLS_REQUIRED=true
LLDAP_SMTP_OPTIONS__SMTP_ENCRYPTION=TLS
```

This is surprising to me, I don't understand what either one is doing on its own. I would love to see docs explaining this; notably, the first option (tls_required) does NOT appear in the [sample config template](https://github.com/lldap/lldap/blob/main/lldap_config.docker_template.toml).

I will also note that once I had the correct settings, I started to see SMTP stuff appearing in the log! But it would have been helpful to see literally anything there when it _wasn't_ working correctly :)


Here's the winning configuration:

```
Jul 30 18:17:29 lldap lldap[5757]:     smtp_options: MailOptions {
Jul 30 18:17:29 lldap lldap[5757]:         enable_password_reset: true,
Jul 30 18:17:29 lldap lldap[5757]:         from: Some(
Jul 30 18:17:29 lldap lldap[5757]:             Mailbox {
Jul 30 18:17:29 lldap lldap[5757]:                 name: None,
Jul 30 18:17:29 lldap lldap[5757]:                 email: Address {
Jul 30 18:17:29 lldap lldap[5757]:                     serialized: "myname@example.com",
Jul 30 18:17:29 lldap lldap[5757]:                     at_start: 8,
Jul 30 18:17:29 lldap lldap[5757]:                 },
Jul 30 18:17:29 lldap lldap[5757]:             },
Jul 30 18:17:29 lldap lldap[5757]:         ),
Jul 30 18:17:29 lldap lldap[5757]:         reply_to: None,
Jul 30 18:17:29 lldap lldap[5757]:         server: "smtp.fastmail.com",
Jul 30 18:17:29 lldap lldap[5757]:         port: 465,
Jul 30 18:17:29 lldap lldap[5757]:         user: "myname@example.com",
Jul 30 18:17:29 lldap lldap[5757]:         password: ***SECRET***,
Jul 30 18:17:29 lldap lldap[5757]:         smtp_encryption: Tls,
Jul 30 18:17:29 lldap lldap[5757]:         tls_required: Some(
Jul 30 18:17:29 lldap lldap[5757]:             true,
Jul 30 18:17:29 lldap lldap[5757]:         ),
Jul 30 18:17:29 lldap lldap[5757]:     },
```

**Reply by nitnelave** [reply]
You forgot to mention this line in the output: 
"DEPRECATED: smtp_options.tls_required field is deprecated, it never did anything. You can replace it with smtp_options.smtp_encryption"
:)

That's why it's not documented. It's just there for backward compatibility. 

There are errors printed to the log if the SMTP settings are wrong and you can't send an email. However, there are no errors if for some reason the password reset is not enabled (it will silently do nothing), which is why I wanted to have a look at the _parsed_ config.

**Reply by corydodt** [reply]
I've unmarked my answer because restarting the server with the same settings resulted in it no longer working. I don't have a clue what's going on any more, but the settings I'm currently using are posted up there. With these settings, there is no log output after I click the password reset button other than the request itself. As you can see, enable_password_reset is on.

Any suggestions on getting better log output or a troubleshooting tool?

**Reply by nitnelave** [reply, marked as answer]
Have you tried with just `lldap send_test_email --to=your@address --verbose`

**Reply by corydodt** [reply]
This works fine and there are no errors. Thanks!

(In my container this was: `sudo podman exec -it lldap /app/lldap send_test_email --to=your@address --verbose`)

Does lldap have some kind of delay period where you can't resend the password email? The way it's behaving it seems like it will send a reset email as long as it hasn't sent one too recently for the same account. If that is what's happening, it kind of explains everything.

(Marking this as the answer to my original question of how to troubleshoot.)

**Reply by nitnelave** [reply]
No, there's no cooldown, or spam protections in general.

**Reply by zachwarner** [reply]
Sorry to glom onto this and revive an old thread, but I'm testing something similar today and seeing weird behavior. First, thanks @corydodt, as
```shellscript
sudo docker exec -it lldap /app/lldap send_test_email --to=your@address --verbose
```
was very helpful. 

However, in my case, I get a `Connection refused` error because, per debug logs, the `Configuration` being used is the default `lldap_config.docker-compose.toml` and not the config I created the container with. I can verify from the `docker compose up` that the correct config is used for initialization, and if I `docker cp lldap:/data/lldap_config.toml /some/host/path/cfg_in_container.toml && nvim /some/host/path/cfg_in_container.toml` I can verify that the correct cfg exists inside the container. So what gives?

**Reply by nitnelave** [reply]
So, in the verbose logs of your container (setting LLDAP_VERBOSE=true), you should see at the top a "Loading configuration from " and the file that gets loaded. 
And then you see the final resolved config as parsed, with environment variables applied and everything. Can you check that the path is as expected? 

If it's not "lldap_config.toml", it means that either: 
- you pass a command line argument --config_file=...
- you set LLDAP_CONFIG_FILE=...
These are the only 2 possibilities. Check your docker compose, your env, etc.

**Reply by zachwarner** [reply]
Apologies, let me provide more information to better diagnose.

### toml

At `/opt/docker/lldap/data/lldap_config.toml`:

```toml
verbose=true
ldap_host = "0.0.0.0"
ldap_port = 3890
http_host = "0.0.0.0"
http_port = 17170
http_url = "https://ldap.EXAMPLE.uk"
ldap_base_dn = "dc=EXAMPLE,dc=uk"
ldap_user_dn = "admin"
ldap_user_email = "noreply@EXAMPLE.uk"
database_url = "sqlite:///data/users.db?mode=rwc"
key_seed = "SOME_RANDOM_STRING"
key_file = '' # required to ensure we use the key_seed

[smtp_options]
server="smtp.gmail.com"
port=587
smtp_encryption="TLS"
user="MY_EMAIL@gmail.com"
password="GMAIL PASSKEY"
from="[LLDAP] <noreply@EXAMPLE.uk>"
reply_to="Do not reply <noreply@EXAMPLE.uk>"

[ldaps_options]
enabled=false

```

### compose

Lives at `/Documents/hms/docker-compose.yml`:

```dockercompose
name: hms
secrets:
    LDAP_PASSWORD:
        file: /opt/docker/authelia/secrets/LDAP_PASSWORD
    LLDAP_JWT_SECRET:
        file: /opt/docker/lldap/secrets/LLDAP_JWT_SECRET
include:
    - ./services/lldap.yml
```

### lldap service

Lives at `/Documents/hms/services/lldap.yml`:

```yml
# LDAP authentication server
services:
    lldap:
        image: lldap/lldap:stable
        container_name: lldap
        secrets:
            - LLDAP_JWT_SECRET
            - LDAP_PASSWORD
        environment:
            - UID=1001
            - GID=1002
            - TZ=America/Chicago
            - LLDAP_JWT_SECRET_FILE=/run/secrets/LLDAP_JWT_SECRET
            - LLDAP_LDAP_USER_PASS_FILE=/run/secrets/LDAP_PASSWORD
        volumes:
            - /opt/docker/lldap/data:/data
        ports:
            - 3890:3890 # turned on during troubleshooting
            - 17170:17170
        restart: unless-stopped
```

### launched with

`bash`, fresh Ubuntu 24 install:

```shellscript
docker compose -f /Documents/hms/docker-compose.yml up --remove-orphans
```

### logs on startup

```log
Attaching to lldap
lldap      | > Setup permissions..
lldap      | > Starting lldap..
lldap      | 
lldap      | Loading configuration from /data/lldap_config.toml
lldap      | Configuration: Configuration {
lldap      |     ldap_host: "0.0.0.0",
lldap      |     ldap_port: 3890,
lldap      |     http_host: "0.0.0.0",
lldap      |     http_port: 17170,
lldap      |     jwt_secret: ***SECRET***,
lldap      |     ldap_base_dn: "dc=EXAMPLE,dc=uk",
lldap      |     ldap_user_dn: "admin",
lldap      |     ldap_user_email: "noreply@EXAMPLE.uk",
lldap      |     ldap_user_pass: ***SECRET***,
lldap      |     force_ldap_user_pass_reset: False,
lldap      |     force_update_private_key: false,
lldap      |     database_url: "sqlite:///data/users.db?mode=rwc",
lldap      |     ignored_user_attributes: [],
lldap      |     ignored_group_attributes: [],
lldap      |     verbose: true,
lldap      |     key_file: "",
lldap      |     key_seed: Some(
lldap      |         ***SECRET***,
lldap      |     ),
lldap      |     smtp_options: MailOptions {
lldap      |         enable_password_reset: false,
lldap      |         from: Some(
lldap      |             ""[LLDAP]" <noreply@EXAMPLE.uk>",
lldap      |         ),
lldap      |         reply_to: Some(
lldap      |             "Do not reply <noreply@EXAMPLE.uk>",
lldap      |         ),
lldap      |         server: "smtp.gmail.com",
lldap      |         port: 587,
lldap      |         user: "MY_EMAIL@gmail.com",
lldap      |         password: ***SECRET***,
lldap      |         smtp_encryption: Tls,
lldap      |         ..
lldap      |     },
lldap      |     ldaps_options: LdapsOptions {
lldap      |         enabled: false,
lldap      |         port: 6360,
lldap      |         cert_file: "cert.pem",
lldap      |         key_file: "key.pem",
lldap      |     },
lldap      |     http_url: "https://ldap.EXAMPLE.uk/",
lldap      |     ..
lldap      | }
lldap      | Generating the private key from the key_seed
lldap      | 2024-12-04T21:41:53.426148410+00:00  INFO     set_up_server [ 5.14ms | 19.94% / 100.00% ]
lldap      | 2024-12-04T21:41:53.426230733+00:00  INFO     ┝━ ｉ [info]: Starting LLDAP version 0.6.1
lldap      | 2024-12-04T21:41:53.427983681+00:00  DEBUG    ┝━ get_schema_version [ 1.28ms | 24.93% ]
lldap      | 2024-12-04T21:41:53.429576600+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | return: Some(SchemaVersion(10))
lldap      | 2024-12-04T21:41:53.430762953+00:00  DEBUG    ┝━ list_groups [ 1.28ms | 24.96% ] filters: Some(DisplayName("lldap_admin"))
lldap      | 2024-12-04T21:41:53.432516826+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | return: [Group { id: 1, display_name: "lldap_admin", creation_date: 2024-12-04T20:50:23.607134739, uuid: "f624bf95-9e06-3da9-a0d2-928f91672a28", users: ["admin", "zach"], attributes: [] }]
lldap      | 2024-12-04T21:41:53.432521239+00:00  DEBUG    ┝━ list_groups [ 415µs | 8.07% ] filters: Some(DisplayName("lldap_password_manager"))
lldap      | 2024-12-04T21:41:53.433082860+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | return: [Group { id: 2, display_name: "lldap_password_manager", creation_date: 2024-12-04T20:50:23.614926464, uuid: "959e34a8-1136-389b-8e4d-ebb60155057f", users: [], attributes: [] }]
lldap      | 2024-12-04T21:41:53.433086603+00:00  DEBUG    ┝━ list_groups [ 370µs | 7.21% ] filters: Some(DisplayName("lldap_strict_readonly"))
lldap      | 2024-12-04T21:41:53.433540468+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | return: [Group { id: 3, display_name: "lldap_strict_readonly", creation_date: 2024-12-04T20:50:23.621016677, uuid: "0dca5848-cf22-36f7-b3b6-7ff915bd91b1", users: [], attributes: [] }]
lldap      | 2024-12-04T21:41:53.433571681+00:00  DEBUG    ┝━ list_users [ 613µs | 11.94% ] filters: Some(MemberOf("lldap_admin")) | _get_groups: false
lldap      | 2024-12-04T21:41:53.434656880+00:00  DEBUG    │  ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: "admin", email: "noreply@EXAMPLE.uk", display_name: Some("Administrator"), creation_date: 2024-12-04T20:50:23.629843739, uuid: "7afdbb9a-5f1c-3f8f-b0d6-cedee4e6db0e", attributes: [] }, groups: Some([GroupDetails { group_id: 1, display_name: "lldap_admin", creation_date: 2024-12-04T20:50:23.607134739, uuid: "f624bf95-9e06-3da9-a0d2-928f91672a28", attributes: [] }]) }, UserAndGroups { user: User { user_id: "zach", email: "MY_EMAIL@gmail.com", display_name: Some("Zach"), creation_date: 2024-12-04T20:51:04.061777709, uuid: "2c6b8050-1729-3313-ac35-907a80a42641", attributes: [AttributeValue { name: AttributeName(CaseInsensitiveString("first_name")), value: Serialized("Zach") }, AttributeValue { name: AttributeName(CaseInsensitiveString("last_name")), value: Serialized("Warner") }] }, groups: Some([GroupDetails { group_id: 1, display_name: "lldap_admin", creation_date: 2024-12-04T20:50:23.607134739, uuid: "f624bf95-9e06-3da9-a0d2-928f91672a28", attributes: [] }]) }]
lldap      | 2024-12-04T21:41:53.434848130+00:00  INFO     ┝━ ｉ [info]: Starting the LDAP server on port 3890
lldap      | 2024-12-04T21:41:53.435002970+00:00  DEBUG    ┝━ get_jwt_blacklist [ 152µs | 2.96% ]
lldap      | 2024-12-04T21:41:53.435203344+00:00  INFO     ┕━ ｉ [info]: Starting the API/web server on port 17170
lldap      | 2024-12-04T21:41:53.435518119+00:00  INFO     ｉ [info]: starting 1 workers
lldap      | 2024-12-04T21:41:53.435550278+00:00  INFO     ｉ [info]: Actix runtime found; starting in Actix runtime
lldap      | 2024-12-04T21:41:53.435575022+00:00  INFO     ｉ [info]: starting service: "ldap", workers: 1, listening on: 0.0.0.0:3890
lldap      | 2024-12-04T21:41:53.435579227+00:00  INFO     ｉ [info]: starting service: "http", workers: 1, listening on: 0.0.0.0:17170
lldap      | 2024-12-04T21:41:53.437279095+00:00  INFO     ｉ [info]: DB Cleanup Cron started
lldap      | 2024-12-04T21:42:23.249376054+00:00  INFO     ｉ [info]: LDAP session start: a3c8b6b6-27d2-4f4c-a6c2-e16ffdb80dd9
lldap      | 2024-12-04T21:42:23.249712768+00:00  INFO     LDAP request [ 229µs | 100.00% ] session_id: a3c8b6b6-27d2-4f4c-a6c2-e16ffdb80dd9
lldap      | 2024-12-04T21:42:23.249760716+00:00  DEBUG    ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 0, op: SearchRequest(LdapSearchRequest { base: "", scope: Base, aliases: Never, sizelimit: 0, timelimit: 0, typesonly: false, filter: Present("objectClass"), attrs: ["supportedExtension"] }), ctrl: [] }
lldap      | 2024-12-04T21:42:23.249784506+00:00  DEBUG    ┝━ 🐛 [debug]: rootDSE request
...
[rootDSE requests every 30s]
```

### logs on `send_test_email`

Invoked with `sudo docker exec -it lldap /app/lldap send_test_email --to=OTHER_EMAIL@gmail.com --verbose`:

```log
Loading configuration from lldap_config.toml
Configuration: Configuration {
    ldap_host: "0.0.0.0",
    ldap_port: 3890,
    http_host: "0.0.0.0",
    http_port: 17170,
    jwt_secret: ***SECRET***,
    ldap_base_dn: "dc=example,dc=com",
    ldap_user_dn: "admin",
    ldap_user_email: "",
    ldap_user_pass: ***SECRET***,
    force_ldap_user_pass_reset: False,
    force_update_private_key: false,
    database_url: "sqlite://users.db?mode=rwc",
    ignored_user_attributes: [],
    ignored_group_attributes: [],
    verbose: true,
    key_file: "server_key",
    key_seed: None,
    smtp_options: MailOptions {
        enable_password_reset: false,
        from: None,
        reply_to: None,
        server: "localhost",
        port: 587,
        user: "",
        password: ***SECRET***,
        smtp_encryption: Tls,
        ..
    },
    ldaps_options: LdapsOptions {
        enabled: false,
        port: 6360,
        cert_file: "cert.pem",
        key_file: "key.pem",
    },
    http_url: "http://localhost/",
    ..
}
2024-12-04T22:02:05.908836225+00:00  DEBUG    🐛 [debug]: Sending email to 'OTHER_EMAIL@gmail.com' as 'LLDAP <nobody@lldap>' via ''@'localhost':'587'
2024-12-04T22:02:05.909455678+00:00  DEBUG    🐛 [debug]: Error sending email: lettre::transport::smtp::Error { kind: Connection, source: Os { code: 111, kind: ConnectionRefused, message: "Connection refused" } }
Error: Could not send email: {:#}

Caused by:
    Error sending email: Connection error: Connection refused (os error 111)
```

### discussion

It looks like it _is_ loading from `lldap_config.toml` (per the log line in `send_test_email`), but many fields do not seem to be persisted (`ldap_base_dn`, `ldap_user_email`, `key_file`, `key_seed`, `http_url`, and all of the `smtp_options`.) 

Edited to add: in case it isn't obvious, I've done some anonymization here (in all caps), but the logs/configs reflect the correct values for my domain/email/etc.

**Reply by nitnelave** [reply]
The config seems to be loading fine for the main service. It's when you run the send email command that the values are wrong. You need to either run it from the /data directory, or pass it an explicit --config_fil2e=/data/lldap_config.toml

**Reply by zachwarner** [reply]
Thanks @nitnelave! I didn't realize I'd need to manually pass the config file. For posterity,

```shellscript
sudo docker exec -it lldap /app/lldap send_test_email --to=OTHER_EMAIL@gmail.com --verbose --config-file=/data/lldap_config.toml
```

is the magic incantation.

For anyone else who stumbles on this looking for a gmail example, it also required `smtp_encryption="STARTTLS"` (not `"TLS"`) in `lldap_config.toml`.

**Reply by zachwarner** [reply]
Apologies but one more: is there a corresponding cli arg to use the SMTP password that's set via secret? I poked around the repo a bit but couldn't find anything. Thanks!

**Reply by nitnelave** [reply]
No, there isn't. If you have a shell, you can do `--smtp-password=$(cat myfile)`
